### PR TITLE
Implement RFC 1679

### PR DESCRIPTION
**Pull request by sfackler**
cc #35729

r? @alexcrichton

### CODE REVIEW

**Comment by alexcrichton** [comment]
Nice! Could you also add tests that are exercising these new abilities as well?

**Comment by sfackler** [comment]
@alexcrichton updated

**Comment by alexcrichton** [comment]
Ok so I finally got around to actually running this on crater, and the result was three regressions: https://gist.github.com/alexcrichton/0c44e2a95044ae9f4d364496bbd0a388. One of those is spurious but the other two look legitimate.

* [The first](https://github.com/jameshurst/rss-rs/blob/master/src/fromxml.rs#L171-L175) has some weird lifetime error
* [The second](https://gitlab.com/Arvamer/gilrs/blob/master/src/mapping.rs#L209) looks like some type inference failures

These seem kinda worrisome :(

**Comment by sfackler** [comment]
The gilrs failure looks like the kind of breakage we'd expect from our back compat guarantees and should be easy to fix. The rss-rs issue is way weirder - not sure what's going on with that.

**Comment by alexcrichton** [comment]
I'm guessing this is all because the output type is now generic, so there's cases where there's just not enough inference to figure out what it should actually be

**Comment by alexcrichton** [comment]
Ok, minimizing the rss-rs case a little further:

```rust
fn foo<'a>(split: &[&'a [u8]]) -> (&'a [u8], &'a [u8]) {
    (
        unsafe { split.get_unchecked(0) },
        unsafe { split.get_unchecked(1) },
    )
}
```

On stable that compiles and with this patch it yields

```
error[E0495]: cannot infer an appropriate lifetime for autoref due to conflicting requirements
 --> src/lib.rs:3:24
  |
3 |         unsafe { split.get_unchecked(0) },
  |                        ^^^^^^^^^^^^^

error[E0495]: cannot infer an appropriate lifetime for autoref due to conflicting requirements
 --> src/lib.rs:4:24
  |
4 |         unsafe { split.get_unchecked(1) },
  |                        ^^^^^^^^^^^^^
  |
help: consider using an explicit lifetime parameter as shown: fn foo<'a>(split: &'a [&'a [u8]]) -> (&'a [u8], &'a [u8])
 --> src/lib.rs:1:1
  |
1 | fn foo<'a>(split: &[&'a [u8]]) -> (&'a [u8], &'a [u8]) {
  | ^

error: aborting due to 2 previous errors
```

That help message seems pretty suspicious for what might be happening here/

**Comment by sfackler** [comment]
@alexcrichton do you see the same error using normal indexing with/without the patch?

**Comment by alexcrichton** [comment]
@sfackler interestingly, no! If I change the return value to `(split[0], split[1])`, it works on stable, nightly, and with this patch. Even if I change it to `(&split[0], &split[1])` it works on all three compilers...

**Comment by sfackler** [comment]
That's bizarre. It is a difference between parameterized impls and parameterized methods, but I wouldn't have expected that to matter in this way.

**Comment by aturon** [comment]
cc @nikomatsakis, see https://github.com/rust-lang/rust/pull/36340#issuecomment-246517485 in particular.

**Comment by nikomatsakis** [comment]
Hmm. I have some theories but no concrete idea. I think I'll have to dig in a bit.

**Comment by nikomatsakis** [comment]
I spent about an hour digging into this case today. Got to do a few other things now, but I'll try to come back to it. I haven't yet gotten to the bottom of what's going on, but I am going to leave myself a few notes here:

- the value being returned has type `&'1 &'2 [u8]`, which is then automatically coerced via deref-coercion to `&'a [u8]` (those are region variables). IOW, the compiler inserts to something like `&**split.get_unchecked(0)`. However, if you add this by hand, everything works. =) So that's a sign of a compiler bug.
- the bad constraint seems to be `'a <= '1`, which is added as part of a subtype check when subtyping `&'1 [u8] <: &'a [u8]`. That is ... valid but I'm not yet quite sure why this subtyping is happening. Anyway, I *think* this is the problem.
- here `'1` represents the lifetime of the return value; there is a constraint that `'1 <= '0`, where `'0` is the lifetime of the reference to `self` that is passed into `get_unchecked`. This seems correct too.

**Comment by sfackler** [comment]
Thanks for looking into it!

@alexcrichton Given that the weird breakage seems to be something that "should" work modulo compiler bugs, I'd lean towards fixing the two crates that regress and landing this. Thoughts?

**Comment by alexcrichton** [comment]
Hm this is a change to such a core type I'd personally prefer to wait for the compiler bug, if any, to get fixed.

**Comment by nikomatsakis** [comment]
I'd be more comfortable waiting; I'll try to dig in a bit more asap

**Comment by nikomatsakis** [comment]
OK, sorry for the delay, last week was crazy, but I dug in some more and I understand the problem. In particular, we have this bit of cleverness called `expected_types_for_fn_args()` which attempts to propagate constraints from the return type to the arguments. The intention here was things like `Some(foo)` apply coercions to `foo`. 

In this case, though, we know that the return type like be `&'1 [T]`, and we have the expected return type of `&'a [u8]`. From this we conclude (falsely) that we should give an expected type of `&'a [u8]` and we relate that `'1: 'a`. But ultimately the coercion becomes `&**` which doesn't require that `'1: 'a`.

I'm not entirely sure what's the best solution here. I have to kind of bring this coercion bit back in my head, and in particular under what circumstances it is required to commit region edges; it may be that we can apply it in more limited circumstances, or avoid committing all of the region inference edges. I suspect region edges are only needed when they affect new region variables that get created during the coercion -- e.g., the edge here, if it were truly needed, would be recreated, I believe.

cc @eddyb who wrote the code originally

**Comment by eddyb** [comment]
Yeah, the region stuff was never great. We only truly care about "type skeletons" - as long as the coercion is triggered, having more freedom around regions (i.e. even replacing them all with fresh vars) shouldn't break anything.
Regular subtyping and then later regionck should (re)create all necessary relationships anyway.

**Comment by nikomatsakis** [comment]
@eddyb 

>  Regular subtyping and then later regionck should (re)create all necessary relationships anyway.

Do you remember why we had to commit the regions? I guess I can remove that step (and perhaps replace them with fresh vars) and see what happens.

**Comment by nikomatsakis** [comment]
OK, I have a provisional fix for the compiler bug in question. Doing more testing now.

**Comment by brson** [comment]
Thanks @nikomatsakis!

**Comment by sfackler** [comment]
Rebased - can we run crater one more time now that niko's fix is in?

**Comment by eddyb** [comment]
Starting crater run.

**Comment by sfackler** [comment]
Thanks!

File: src/test/compile-fail/on-unimplemented/slice-index.rs [hunk 1]
```diff
@@ -17,12 +17,12 @@ use std::ops::Index;
...
-    //~^ ERROR E0277
-    //~| NOTE the trait `std::ops::Index<std::ops::RangeTo<i32>>` is not implemented for `[i32]`
-    //~| NOTE slice indices are of type `usize`
+    x[1i32]; //~ ERROR E0277
+             //~| NOTE slice indices are of type `usize` or ranges of `usize`
+             //~| NOTE trait `std::slice::SliceIndex<i32>` is not implemented for `i32`
+             //~| NOTE required because of the requirements on the impl of `std::ops::Index<i32>`
+    x[..1i32]; //~ ERROR E0277
+               //~| NOTE slice indices are of type `usize` or ranges of `usize`
+               //~| NOTE trait `std::slice::SliceIndex<i32>` is not implemented for `std::ops::RangeTo<i32>`
```
Comment (eddyb) [inline src/test/compile-fail/on-unimplemented/slice-index.rs]: tidy failed, `line longer than 100 chars` (should not impact the crater run though).

Comment (sfackler) [inline src/test/compile-fail/on-unimplemented/slice-index.rs]: Oops, fixed.

**Comment by eddyb** [comment]
[Crater report](https://gist.github.com/eddyb/5d1eb5cd83a3129168433b3d8f24a97a) shows one root regression (`postgres`).

**Comment by sfackler** [comment]
Well that's easy! https://github.com/sfackler/rust-postgres/commit/11f11864391ae457881a0377f89b154e897d094d

**Comment by sfackler** [comment]
@alexcrichton I think this is good to go now?

File: src/test/run-fail/bounds-check-no-overflow.rs [hunk 2]
```diff
@@ -8,7 +8,7 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-// error-pattern:index out of bounds
+// error-pattern:assertion failed: self < slice.len()
```
Comment (frewsxcv) [inline src/test/run-fail/bounds-check-no-overflow.rs]: As per travis:

```
error: error pattern 'assertion failed: self < slice.len()' not found!
```

Comment (sfackler) [inline src/test/run-fail/bounds-check-no-overflow.rs]: Oops, fixed.

**Comment by alexcrichton** [comment]
@bors: r+

Nice!

**Pull request merged**